compiler/src/iree/compiler/Codegen/Dialect/Codegen/Utils/Utils.cpp

```
@@ -284,7 +284,11 @@ getExpandedTileShape(const TileSwizzle::ExpandShapeType &expandShape) {
 MaterializeEncodingInfo
 getEncodingInfoForMatmul(Encoding::EncodingAttr encoding, TileMxNxK tileMxNxK) {
   MaterializeEncodingInfo encodingInfo;
-  auto cDims = getEncodingContractionDims(encoding);
+  FailureOr<linalg::ContractionDimensions> cDims =
```

This can for example result in a failure if `user_indexing_maps` isn't provided, hitting the assert below. The following test checks this case: https://github.com/iree-org/iree/pull/20027/files#diff-1308306ae2802ec5bb022c31338cea3cd8dac61a9cb675541c9bd37009161e16R84

---

My guess is that this is not needed if you remove other tests. We will need new features when we consider layout transfers. For now, it is fine because most of cases are single device today

---

This would be good to have in any case I think, an IR change shouldn't trigger an assert? If this should never happen, we should instead throw an error, but I think returning an empty `MaterializeEncodingInfo` would be better.

---

I think what you said is reasonable, thanks! I had two thoughts, and TLDR is that I think we can keep it this way for now. :)

1. In the past, we do not have the pre-resolved layouts information in encodings. My mental model was that it is always for devices to resolve high level encoding information into physical encoding information. This is why there was a hidden assumption that it'd always succeed. According to the method name and usage, it is okay to return an identity materialize encoding info here. So I think your change makes sense now.
2. The second thought is that the `encoding` has `layouts` field. They can resolve some information from `layouts`, if any. In this context, they can return deserialized encoding info. However, I'm considering if we should materialize encoding_attr to other encoding attribute that has DictionaryAttr only. I don't know yet, so I think we can go with (1) (i.e., what you've done) for now.

---

This is related to having a pre-resolved layouts, but also orthogonal. I encountered the issue with pre-resolved layouts, but it can be hit in different ways if certain information is not available (like `user_indexing_maps`), so I think returning an identity `MaterializeEncodingInfo` is good behaviour as not all needed information is available (possibly caused by a pre-resolved layout).

For pre-resolved layouts, I was wondering about 2) as well, and whether you could resolve the encoding info from that, but it wasn't clear to me to what extent. Like, if we're going to recreate the same layout, there is no point and we could just skip resolving the encoding from the layout.

---

RE pre-resolved layouts, I took your idea and addressed it in https://github.com/iree-org/iree/pull/20037